#!/bin/bash

#PBS -l nodes=1:ppn=8,walltime=40:00:00
#PBS -N gatk_geno_GVCFs
#PBS -m abe
#PBS -M cxb585@psu.edu
#PBS -e localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.e${PBS_JOBID}
#PBS -o localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.o${PBS_JOBID}

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

working_dir=$PBS_O_WORKDIR

module load gatk

GENOME_FA=genomes/rheMac2/rheMac2.fa
GENOME_NAME=rhesus

# ------------------------------------------------------------------------------
# Run program
# ------------------------------------------------------------------------------

cd $working_dir

GVCFS=(`ls results/*.PE.bwa.${GENOME_NAME}.passed.realn.g.vcf`)

count=0
for v in ${GVCFS[*]}; do
	GVCFS[$count]="--variant $v "
	count=`expr $count + 1`
done

# Make output directory
mkdir -p ${GENOME_NAME}_snps

java -jar ${GATK}/GenomeAnalysisTK.jar \
    -T GenotypeGVCFs \
    -R $GENOME_FA \
    ${GVCFS[*]} \
    -o ${GENOME_NAME}_snps/ALL.raw.snps.indels.vcf \
    -nt 12

exit;
