#!/bin/bash

#$ -N gatk_genotype
#$ -M cbergey@nd.edu
#$ -m abe
#$ -q *@@bio
#$ -pe smp 8

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

working_dir=$SGE_O_WORKDIR

# ------------------------------------------------------------------------------
# Run pipeline
# ------------------------------------------------------------------------------

cd $working_dir

module load bio/gatk/3.3 

#module load jdk/1.7.0

GATK=/afs/crc.nd.edu/x86_64_linux/bio/gatk/3.3
GENOME_FA=genomes/AgamP4/AgamP4.fa
GENOME_NAME=AgamP4

# ------------------------------------------------------------------------------
# Run program
# ------------------------------------------------------------------------------

cd $working_dir

CHROM=$(cut -f1 genomes/AgamP4/AgamP4.fa.fai | head -n $SGE_TASK_ID | tail -n 1)

BAMS=(`ls results/*.PE.bwa.${GENOME_NAME}.passed.realn.bam`)

count=0
for b in ${BAMS[*]}; do
	BAMS[$count]="-I "$b" "
	count=`expr $count + 1`
done

# Make output directory
mkdir ${GENOME_NAME}_snps

java -jar ${GATK}/GenomeAnalysisTK.jar \
	-T UnifiedGenotyper \
	-R ${GENOME_FA} \
	${BAMS[*]} \
	-stand_call_conf 50.0 \
	-stand_emit_conf 10.0 \
	-o ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.vcf \
	-nct 4 \
	-nt 8 \
	-L ${CHROM}

exit;
