#!/bin/bash

#$ -N filter_gatk_snps
#$ -M cbergey@nd.edu
#$ -m abe
#$ -q *@@bio
#$ -pe smp 8

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

working_dir=$SGE_O_WORKDIR

module load bio/gatk/3.3 

#module load jdk/1.7.0

GATK=/afs/crc.nd.edu/x86_64_linux/bio/gatk/3.3
GENOME_FA=genomes/AgamP4/AgamP4.fa
GENOME_NAME=AgamP4

# ------------------------------------------------------------------------------
# Run program
# ------------------------------------------------------------------------------

cd $working_dir

CHROM=$(cut -f1 genomes/AgamP4/AgamP4.fa.fai | head -n $SGE_TASK_ID | tail -n 1)

java -Xmx2g -jar ${GATK}/GenomeAnalysisTK.jar \
	-R ${GENOME_FA} \
	-T VariantFiltration \
	-o ${GENOME_NAME}_snps/chr${CHROM}.flt.vcf \
	--variant ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.vcf \
	--filterExpression "QD < 2.0" \
	--filterName "QDfilter" \
	--filterExpression "MQ < 40.0" \
	--filterName "MQfilter" \
	--filterExpression "FS > 60.0" \
	--filterName "FSfilter" \
	--filterExpression "HaplotypeScore > 13.0" \
	--filterName "HAPSCfilter" \
	--filterExpression "MQRankSum < -12.5" \
	--filterName "MQRSfilter" \
	--filterExpression "ReadPosRankSum < -8.0" \
	--filterName "RPRSfilter" \
	--missingValuesInExpressionsShouldEvaluateAsFailing	

# Select variants with "FILTER=PASS" and are SNPs:
java -jar ${GATK}/GenomeAnalysisTK.jar \
	-T SelectVariants \
	-R ${GENOME_FA} \
	--variant ${GENOME_NAME}_snps/chr${CHROM}.flt.vcf \
	--select_expressions "vc.isNotFiltered() && vc.isSNP()" \
	-o ${GENOME_NAME}_snps/chr${CHROM}.pass.snp.vcf

exit;
